fathom/src/core/binary.rs

```diff
@@ -484,6 +486,50 @@ impl<'arena, 'env, 'data> Context<'arena, 'env, 'data> {
         }
     }
 
+    fn read_repeat_until_full(
+        &mut self,
+        reader: &mut BufferReader<'data>,
+        len: &ArcValue<'arena>,
+        replicate: &ArcValue<'arena>,
+        elem_format: &ArcValue<'arena>,
+    ) -> Result<ArcValue<'arena>, ReadError<'arena>> {
+        let len = match self.elim_env().force(len).as_ref() {
+            Value::ConstLit(Const::U8(len, _)) => Some(usize::from(*len)),
+            Value::ConstLit(Const::U16(len, _)) => Some(usize::from(*len)),
+            Value::ConstLit(Const::U32(len, _)) => usize::try_from(*len).ok(),
+            Value::ConstLit(Const::U64(len, _)) => usize::try_from(*len).ok(),
+            _ => return Err(ReadError::InvalidValue(len.span())),
+        }
+        .ok_or_else(|| ReadError::InvalidValue(len.span()))?;
+
+        // TODO: Do we need to force replicate as well?
+        // let replicate = self.elim_env().force(replicate);
```

Comment:
Hum. Yeah we might actually. We don’t directly match on the value (which is what would usually make you reach for `force`), but it seems like `ElimEnv::fun_app` doesn’t force the value before matching itself. Not sure if it’s common to force in eliminator helpers like `fun_app`… might need to take a look at the elaboration-zoo.  🤔

Even if `ElimEnv::fun_app` did, because we don’t have mutable flexvars at the moment, it’s probably good for performance to force once before iterating.